### PR TITLE
Run doc snippets twice to check loading from the configuration cache

### PR DESCRIPTION
**Pull request by mlopatkin**
When running the second time, the "--rerun-tasks" is specified to ensure that all doFirst/doLast actions are restored and executed.
The serialization of the Groovy closures used for these skips closure's delegates, so sometimes methods referenced in closures are not resolvable when running from cache.

This is a part of the #21027 work

### CODE REVIEW

**Comment by mlopatkin** [comment]
@bot-gradle test this

**Review by nathan-contino — APPROVED** [review]
LGTM

File: subprojects/docs/src/docsTest/java/org/gradle/docs/samples/IntegrationTestSamplesExecutor.java [hunk 1]
```diff
@@ -62,6 +64,27 @@ class IntegrationTestSamplesExecutor extends CommandExecutor {
...
+                // TODO(mlopatkin) sometimes it is still possible to save the configuration cache state in the event of failure.
+                //  We need to figure out how to separate expected failure from the CC store failure.
+                ExecutionFailure result = createExecuter(args, flags).runWithFailure();
+                outputStream.write((result.getOutput() + result.getError()).getBytes());
+            } else {
+                ExecutionResult result = createExecuter(args, flags).run();
+                outputStream.write(result.getOutput().getBytes());
+
+                if (shouldCheckLoadingFromConfigurationCache()) {
+                    createExecuter(args, flags).withArgument("--rerun-tasks").run();
```
Comment (bamboo) [inline subprojects/docs/src/docsTest/java/org/gradle/docs/samples/IntegrationTestSamplesExecutor.java:78]: ❓ I wonder what happens to the build output when the 2nd run fails 🤔

Comment (mlopatkin) [reply subprojects/docs/src/docsTest/java/org/gradle/docs/samples/IntegrationTestSamplesExecutor.java:78]: It is nicely added to the failure message, see, for example, https://ge.gradle.org/s/2x4hcknqwf43s/tests/:docs:docsTest/org.gradle.docs.samples.Bucket3SnippetsTest/snippet-ear-ear-with-war_groovy_earWithWar.sample?top-execution=1

File: subprojects/docs/src/docsTest/java/org/gradle/docs/samples/IntegrationTestSamplesExecutor.java [hunk 2]
```diff
@@ -62,6 +65,40 @@ class IntegrationTestSamplesExecutor extends CommandExecutor {
...
+        }
+    }
+
+    private void rerunReusingStoredConfigurationCacheEntry(List<String> args, List<String> flags) {
+        // Rerun tasks. If the task is up-to-date, then its actions aren't executed.
+        // This might hide issues with Groovy closures, as the methods referenced inside the closure are resolved dynamically.
+        try {
+            createExecuter(args, flags).withArgument("--rerun-tasks").run();
+        } catch (UnexpectedBuildFailure e) {
+            UnexpectedBuildFailure tweakedException = new UnexpectedBuildFailure("Failed to rerun the build when reusing the configuration cache entry.\n" + e.getMessage());
```
Comment (mlopatkin) [inline subprojects/docs/src/docsTest/java/org/gradle/docs/samples/IntegrationTestSamplesExecutor.java:95]: A little trick to provide a better error message, while printing the single original stack. A typical causal chain is not very suitable, as it either pushes the interesting Gradle output down to `caused by: ...` section or forces us to print the output twice. With this, we can have output like [there](https://ge.gradle.org/s/koesyiw2jzcv6/tests/:docs:docsTest/org.gradle.docs.samples.Bucket2SnippetsTest/snippet-tasks-incremental-task_groovy_incrementalTaskChangedProperty.sample?top-execution=1). Adding the original exception as a cause actually [looks acceptable in GE](https://ge.gradle.org/s/cncnkjnb5xvla/tests/:docs:docsTest/org.gradle.docs.samples.Bucket2SnippetsTest/snippet-tasks-incremental-task_groovy_incrementalTaskChangedProperty.sample?top-execution=1), but the stacktrace would be unfolded everywhere else.

**Comment by mlopatkin** [comment]
@bot-gradle test and merge

**Pull request merged**